File: lintcode/longest_substring_without_repeating_characters.py
class Solution:
    # @param s: a string
    # @return: an integer
    def lengthOfLongestSubstring(self, s):
        begin, max_length, last_index = 0, 0, {}
        for index, c in enumerate(s):
            if c in last_index and begin < last_index[c] + 1:
                begin = last_index[c] + 1
            else:
                max_length = max(max_length, index - begin + 1)
            last_index[c] = index
        return max_length

File: leetcode/longest_substring_without_repeating_characters.js
/* Given a string, find the length of the longest substring without repeating 
 * characters. For example, the longest substring without repeating letters 
 * for "abcabcbb" is "abc", which the length is 3. 
 * For "bbbbb" the longest substring is "b", with the length of 1.
 * Time Complexity: O(n) in average
 * Space Complexity: O(n)
 */

/**
 * @param {string} s
 * @return {number}
 */
var lengthOfLongestSubstring = function(s) {
    var max_length = 0, last = {};
    for (var begin = 0, i = 0; i < s.length; ++i ) {
        if (s[i] in last && begin < last[s[i]] + 1)
            begin = last[s[i]] + 1;
        else
            max_length = Math.max(max_length, i - begin + 1);
        last[s[i]] = i;    
    }
    return max_length;
};
